# HiTaxon: A hierarchical ensemble framework for taxonomic classification of short reads

## Overview

**HiTaxon** is an automated framework for creating custom short-read taxonomic classifiers for various environments. Given a list of genera, HiTaxon downloads and processes assemblies from RefSeq such that it creates a set of non-redundant sequences for species encompassed in the genera. HiTaxon then builds custom databases for a reference-dependent classifier and uses the classifier to generate taxonomic predictions on input FASTA files. The reference dependent genus-level outputs are used to determine which HiTaxon trained FastText machine learning classifiers are used for species predictions.

As per the results of our paper, Kraken2 ensembles provided the best performance and is consequently the reference-dependent classifier of choice for HiTaxon. However code for Kaiju is available under other_classifiers/kaiju_classifiers

Please refer to our paper for more information:

> placeholder

Data used to evaluate classifiers in our study can be accessed at: 

https://doi.org/10.5281/zenodo.8335901


## Installation

Clone the repository
```bash
git clone https://github.com/ParkinsonLab/HiTaxon
```
Navigate to the repository
```bash
cd HiTaxon
```
Create conda environment
```bash
conda env create --name HiTaxon --file environment.yml
```
Activate conda environment 
```bash
conda activate HiTaxon
```
Build HiTaxon
```bash
pip install .
```

## Usage

In order to use HiTaxon, you must create a configuration file using the structure below and save it as config.file in the HiTaxon directory

```
ASSEMBLY_SUMMARY=default
GENUS_NAMES=/path/to/list_of_genera
KRAKEN_NAME=desired_database_name
KRAKEN_PATH=/path/to/store_database
MODEL_PATH=/path/to/store_ML_models
NUM_OF_THREADS=number_of_threads
OUTPUT_PATH=/path/to/download_and_process_RefSeq_data
REPORT_PATH=/path/to/store_taxonomic_predictions
```

If ASSEMBLY_SUMMARY is left as default, HiTaxon will download and use the latest file from NCBI. If you want to use a specific version, provide the path to the file

Note: Only the text file corresponding to GENUS_NAMES needs to be generated by the user. An example of how such a text file is structured is provided below

```
Bacillus
Enterococcus
Escherichia
Lactobacillus
Listeria
Staphylococcus
Salmonella
```

For the remaining directories, HiTaxon will create them if they are not present

### Download High-Quality Sequences

To download sequences from RefSeq corresponding to the predefined genera of interest, use the following command

```bash
./HiTaxon.sh --collect
```

### Remove Redundant Sequences

To process sequences which were downloaded from RefSeq, use the following command

```bash
./HiTaxon.sh --process
```

### Create custom Kraken2 database

To build a database for Kraken2 using the processed collection of non-redundant sequences, use the following command

```bash
./HiTaxon.sh --build
```

### Train ML classifiers

To train genus-specific machine learning classifiers for species prediction, use the following command

```bash
./HiTaxon.sh --train
```

### Generate taxonomic predictions

To generate predictions from metagenomic or metatranscriptomic sequences with a hierarchical ensemble, use the following command

```bash
./HiTaxon.sh --evaluate -f path/to/fasta_file -o name_of_output_report -m Ensemble
```

To reduce runtime for evaluation, we have also added the option to use only the predictions from the custom Kraken2 classifier

```bash
./HiTaxon.sh --evaluate -f path/to/fasta_file -o name_of_output_report -m Kraken2
```

## Links to software used for HiTaxon
- [art](https://anaconda.org/bioconda/art)
- [biopython](https://github.com/biopython/biopython)
- [cd-hit](https://github.com/weizhongli/cdhit)
- [ete](https://github.com/etetoolkit/ete)
- [Fastani](https://github.com/ParBLiSS/FastANI)
- [fastp](https://github.com/OpenGene/fastp)
- [fastq-join](https://github.com/brwnj/fastq-join)
- [fasttext](https://github.com/facebookresearch/fastText)
- [kaiju](https://github.com/bioinformatics-centre/kaiju)
- [kraken2](https://github.com/DerrickWood/kraken2)
- [ncbi datasets](https://github.com/ncbi/datasets)
- [numpy](https://github.com/numpy/numpy)
- [pandas](https://github.com/pandas-dev/pandas)
- [polyester](https://github.com/alyssafrazee/polyester)
- [prodigal](https://github.com/hyattpd/Prodigal)
- [scikit-learn](https://github.com/scikit-learn/scikit-learn)
- [seqkit](https://github.com/shenwei356/seqkit)
- [seqtk](https://github.com/lh3/seqtk)
