# HiTaxon: A hierarchical ensemble framework for taxonomic classification of short reads

## Overview

**HiTaxon** is an automated framework for creating custom short-read taxonomic classifiers for various environments. Given a list of genera, HiTaxon downloads and processes assemblies from RefSeq such that it creates a set of non-redundant sequences for species encompassed in the genera. HiTaxon then creates a custom database for a primary reference-dependent classifier and uses the classifier to generate taxonomic predictions on input FASTA files. The reference dependent genus-level outputs are used to determine which HiTaxon built specialized classifiers for specific genus to set of species pairs are used for species predictions.

Please refer to our paper for more information:

> Verma, B. and Parkinson, J. HiTaxon: A hierarchical ensemble framework for taxonomic classification of short reads. BioArxiv 2023.09.11.557224; https://doi.org/10.1101/2023.09.11.557224

Data used to evaluate classifiers in our study can be accessed at: 

https://doi.org/10.5281/zenodo.8335901


## Installation

Clone the repository
```bash
git clone https://github.com/ParkinsonLab/HiTaxon
```
Navigate to the repository
```bash
cd HiTaxon
```
Create conda environment
```bash
conda env create --name HiTaxon --file environment.yml
```
Activate conda environment 
```bash
conda activate HiTaxon
```
Build HiTaxon
```bash
pip install .
```



## Usage

In order to use HiTaxon, you must create a configuration file using the structure below and save it as config.file in the HiTaxon directory

```
ASSEMBLY_SUMMARY=default
GENUS_NAMES=/path/to/list_of_genera
KRAKEN_NAME=desired_database_name
KRAKEN_PATH=/path/to/store_database
MODEL_PATH=/path/to/store_ML_models
NUM_OF_THREADS=number_of_threads
OUTPUT_PATH=/path/to/download_and_process_RefSeq_data
REPORT_PATH=/path/to/store_taxonomic_predictions
BWA_PATH=/path/to/store_bwa_indices
```

If ASSEMBLY_SUMMARY is left as default, HiTaxon will download and use the latest file from NCBI. If you want to use a specific version, provide the path to the file

Note: Only the text file corresponding to GENUS_NAMES needs to be generated by the user. An example of how such a text file is structured is provided below

```
Bacillus
Enterococcus
Escherichia
Lactobacillus
Listeria
Staphylococcus
Salmonella
```

For the remaining directories, HiTaxon will create them if they are not present

### Download High-Quality Sequences

To download sequences from RefSeq corresponding to the predefined genera of interest, use the following command

```bash
./HiTaxon.sh --collect
```

### Remove Redundant Sequences

To process sequences which were downloaded from RefSeq, use the following command

```bash
./HiTaxon.sh --process
```

### Create Custom Kraken2 Database

To build a database for Kraken2 using the processed collection of non-redundant sequences, use the following command

```bash
./HiTaxon.sh --build
```

### Generate Specialized BWA Indices

To create genus-specific BWA indices for species prediction, use the following command

```bash
./Hitaxon.sh --align
```

### Train Specialized ML Classifiers

To train genus-specific machine learning classifiers for species prediction, use the following command

```bash
./HiTaxon.sh --train
```
Note: You will be prompted by HiTaxon on as to whether the mode is in 1) data_creation  or 2) model_creation, **select option 2** .

### Generate Taxonomic Predictions

To generate predictions for metagenomic or metatranscriptomic reads in a FASTA file using a hierarchical ensemble of Kraken2 and BWA, use the following command

```bash
./HiTaxon.sh --evaluate -f path/to/fasta_file -o name_of_output_report -m Kraken2_BWA
```

To integrate ML classifiers instead of BWA, use the following command

```bash
./HiTaxon.sh --evaluate -f path/to/fasta_file -o name_of_output_report -m Kraken2_ML
```

To reduce runtime for evaluation, we have also added the option to use only the predictions from the custom Kraken2 classifier

```bash
./HiTaxon.sh --evaluate -f path/to/fasta_file -o name_of_output_report -m Kraken2
```

## Use Cases

While we imagine researchers can derive value from HiTaxon through a multiple of use cases, 4 common use cases we might expect are 

1. Building the best taxonomic classifier for a particular dataset when there are moderate time constraints
2. Building the best taxonomic classifier for particular dataset when there are significant time constrains
3. Ensembling existing Kraken2 databases with HiTaxon-curated specialized classifiers
4. Creating Training Data for custom ML classifiers

Consequently, we have provided step-by-step instructions for each potential use case.


### Use Case 1: Building the best taxonomic classifier for a particular dataset (moderate time constraints)

To create the best taxonomic classifier for a particular dataset when there are moderate time constraints, we recommend employing Kraken2-HiTaxon-Align, which is an hierarchical ensemble consisting of a Kraken2 classifier paired with HiTaxon curated sequences and specialized BWA indices. In our manuscript, we highlight that Kraken2-HiTaxon-Align is the best performing taxonomic classifier amongst all that were tested.

In order to build this ensemble, execute the following steps in order:
1. Create a directory to store sequence data from RefSeq.
2. Within this directory, create a file called `taxon.txt` which lists all genera of interest, using the same format highlighted earlier in the documentation.
3. In the HiTaxon directory, create a `config.file` which defines a set of important parameters for HiTaxon, using the same format highlighted earlier in the documentation.
4. Download sequences from RefSeq that pertain to the set of genera listed in `taxon.txt` :
    ```bash
    ./HiTaxon.sh --collect
    ```
5. Cluster similiar assemblies and coding sequences :
    ```bash
    ./HiTaxon.sh --process
    ```
6. Create a custom database composed of non-redundant sequences aquired from Step 5 for Kraken2:
    ```bash
    ./HiTaxon.sh --build
    ```
7. Create specialized BWA indices for each genus to set of species pair using non-redundant sequences aquired from Step 5 for BWA:
    ```bash
    ./HiTaxon.sh --align
    ```

8. Use the hierarchical ensemble to generate taxonomic predictions for short-reads in `input.fasta` :
    ```bash
    ./HiTaxon.sh --evaluate -f path/to/input.fasta -o name_of_output_report -m Kraken2_BWA
    ```

### Use Case 2: Building the best taxonomic classifier for particular dataset (significant time constrains)
To create the best taxonomic classifier for a particular dataset when there are signigicant time constraints, we recommend employing Kraken2-HiTaxon-DB, which is composed entirely of Kraken2 with a HiTaxon curated database. This approach has a small reduction in recall relative to Kraken2-HiTaxon-Align but requires much less time during evaluation.

In order to build this classifier, execute the following steps in order:
1. Create a directory to store sequence data from RefSeq.
2. Within this directory, create a file called `taxon.txt` which lists all genera of interest, using the same format highlighted earlier in the documentation.
3. In the HiTaxon directory, create a `config.file` which defines a set of important parameters for HiTaxon, using the same format highlighted earlier in the documentation.
4. Download sequences from RefSeq pertaining to the set of genera listed in `taxon.txt` :
    ```bash
    ./HiTaxon.sh --collect
    ```
5. Cluster similiar assemblies and coding sequences :
    ```bash
    ./HiTaxon.sh --process
    ```
6. Create a custom database composed of non-redundant sequences aquired from Step 5 for Kraken2:
    ```bash
    ./HiTaxon.sh --build
    ```

7. Use the hierarchical ensemble to generate taxonomic predictions for short-reads in `input.fasta` :
    ```bash
    ./HiTaxon.sh --evaluate -f path/to/input.fasta -o name_of_output_report -m Kraken2
    ```

### Use Case 3: Ensembling existing Kraken2 databases with HiTaxon-curated specialized classifiers
To create a hierarchical ensemble using an existing Kraken2 database with HiTaxon-curated specialized classifiers, execute the following steps in order:

1. Create a directory to store sequence data from RefSeq.
2. Within this directory, create a file called `taxon.txt` which lists all genera of interest, using the same format highlighted earlier in the documentation.
3. In the HiTaxon directory, create a `config.file`, using the same format highlighted earlier in the documentation. **Make sure config.file references the path in which the pre-existing Kraken2 database stored, alongside the name of the database**
4. Download sequences from RefSeq pertaining to the set of genera listed in `taxon.txt` :
    ```bash
    ./HiTaxon.sh --collect
    ```
5. Cluster similiar assemblies and coding sequences :
    ```bash
    ./HiTaxon.sh --process
    ```
6. Option A: Create specialized BWA indices for each genus to set of species pair using non-redundant sequences aquired from Step 5 for BWA:
    ```bash
    ./HiTaxon.sh --align
    ```

6. Option B. Train specialized ML for each genus to set of species pair using non-redundant sequences aquired from Step 5:
   
   Note: You will be prompted by HiTaxon on as to whether the mode is in 1) data_creation  or 2) model_creation, **select option 2** . 
    ```bash
    ./HiTaxon.sh --train
    ```


7. Option A. Use the hierarchical ensemble of Kraken2 and BWA to generate taxonomic predictions for short-reads in `input.fasta` :
    ```bash
    ./HiTaxon.sh --evaluate -f path/to/input.fasta -o name_of_output_report -m Kraken2_BWA
    ```

7. Option B. Use the hierarchical ensemble of Kraken2 and ML classifiers to generate taxonomic predictions for short-reads in  `input.fasta` :
    ```bash
    ./HiTaxon.sh --evaluate -f path/to/input.fasta -o name_of_output_report -m Kraken2_ML
    ```

### Use Case 4: Creating Training Data for custom ML classifiers
To train ML classifiers for HiTaxon, we have built a robust framework for training both multi-class classifiers (i.e genus encompassed multiple species) and binary classifiers (i.e genus encompasses a single species). However, while we used FastText classifiers, we do acknowledge that individual researchers might prefer to test different algorithms for species classification. Consequently, we allow for HiTaxon to create and process training data without forcing the user to also train ML models

To use this feature, execute the following steps in order:


1. Create a directory to store sequence data from RefSeq.
2. Within this directory, create a file called `taxon.txt` which lists all genera of interest, using the same format highlighted earlier in the documentation.
3. In the HiTaxon directory, create a `config.file`, using the same format highlighted earlier in the documentation.
4. Download sequences from RefSeq pertaining to the set of genera listed in `taxon.txt` :
    ```bash
    ./HiTaxon.sh --collect
    ```
5. Cluster similiar assemblies and coding sequences :
    ```bash
    ./HiTaxon.sh --process
    ```
6. Create and preprocess the training data 
    ```bash
    ./HiTaxon.sh --train
    ```
7. You will be promoted whether HiTaxon is in 1) data_creation  or  2) model_creation mode, **select  option 1**. This will generate a .txt for all genus to set of species pairs, for both multi-class and binary problems, which can be easily reformatted by researchers for their ML algorithm of choice.

   Note: If you want to change the K-mer value from the default of K = 13, a single parameter change in train.py, which can be found in scripts/train/, is all that is needed


## Links to software used for HiTaxon
- [art](https://anaconda.org/bioconda/art)
- [biopython](https://github.com/biopython/biopython)
- [bwa](https://github.com/lh3/bwa)
- [cd-hit](https://github.com/weizhongli/cdhit)
- [ete](https://github.com/etetoolkit/ete)
- [Fastani](https://github.com/ParBLiSS/FastANI)
- [fastp](https://github.com/OpenGene/fastp)
- [fastq-join](https://github.com/brwnj/fastq-join)
- [fasttext](https://github.com/facebookresearch/fastText)
- [kaiju](https://github.com/bioinformatics-centre/kaiju)
- [kraken2](https://github.com/DerrickWood/kraken2)
- [ncbi datasets](https://github.com/ncbi/datasets)
- [numpy](https://github.com/numpy/numpy)
- [pandas](https://github.com/pandas-dev/pandas)
- [polyester](https://github.com/alyssafrazee/polyester)
- [prodigal](https://github.com/hyattpd/Prodigal)
- [scikit-learn](https://github.com/scikit-learn/scikit-learn)
- [seqkit](https://github.com/shenwei356/seqkit)
- [seqtk](https://github.com/lh3/seqtk)
